# Introduction to SpaCy - part 2

Let's see how we can convert text to numbers and what we can do with the numbers.

#### Print your name

In [1]:
## Your code here 
print("Exercise by: Heidi Putkuri")

Exercise by: Heidi Putkuri


## Import SpaCy and load pipelines

#### Exercise - Import SpaCy 

- import spacy
- print spacy version
- You may need to download SpaCy text processing pipelines `de_core_news_sm`, `en_core_web_lg` and `fi_core_news_lg`. 

In [2]:
## Task 1:
## Your code here 
import spacy
print("spaCy version: ", spacy.__version__)

!python -m spacy download de_core_news_sm
!python -m spacy download en_core_web_lg
!python -m spacy download fi_core_news_lg


spaCy version:  3.7.2
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.6/14.6 MB 21.5 MB/s eta 0:00:0000:0100:01
✔ Download and installation successful
You can now load the package via spacy.load('de_core_news_sm')
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 587.7/587.7 MB 17.4 MB/s eta 0:00:0000:0100:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.8/230.8 MB 21.0 MB/s eta 0:00:0000:0100:01
✔ Download and installation successful
You can now load the package via spacy.load('fi_core_news_lg')


In [3]:
de_nlp = spacy.load("de_core_news_sm")  
en_nlp = spacy.load("en_core_web_lg")  
fi_nlp = spacy.load("fi_core_news_lg") 

## Strings to hashes

- Look up the string “cat” in `nlp.vocab.strings` to get the hash.
- Look up the hash to get back the string.

In [4]:
import spacy

nlp_en = spacy.load("en_core_web_sm")
nlp_de = spacy.load("de_core_news_sm")
print('English:')
code_en1 = nlp_en.vocab.strings["cat"]
code_en2 = nlp_en.vocab.strings["Katze"]
print('cat = ', code_en1)
print('Katze = ',code_en2)
print('German:')
code_de1 = nlp_de.vocab.strings["cat"]
code_de2 = nlp_de.vocab.strings["Katze"]
print('cat = ', code_de1)
print('Katze = ',code_de2, '\n')
print('English & cat hash:', nlp_en.vocab.strings[code_en1])
print('German & cat hash:', nlp_de.vocab.strings[code_en1])
print('German & Katze hash:', nlp_de.vocab.strings[code_de2])

import gc
del nlp_en
del nlp_de
gc.collect()

/Users/heksaani/miniconda3/envs/ML/lib/python3.9/site-packages/spacy/util.py:910: UserWarning: [W095] Model 'en_core_web_sm' (3.8.0) was trained with spaCy v3.8.0 and may not be 100% compatible with the current version (3.7.2). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate
  warnings.warn(warn_msg)


English:
cat =  5439657043933447811
Katze =  13148695887159701572
German:
cat =  5439657043933447811
Katze =  13148695887159701572 

English & cat hash: cat
German & cat hash: cat
German & Katze hash: Katze


12721

#### Exercise - PERSON vs. Person vs. person hash?
- Use package `en_core_web_sm`
- Create function `str_to_hash(string)` to get hash codes for strings “PERSON / Person / person”.
- Create function `hash_to_str(hash)` to get strings for hash codes.

In [5]:
## Task 2:
## Your code here 
def str_to_hash(string:str):
    nlp_en = spacy.load("en_core_web_sm")
    return nlp_en.vocab.strings[string]



def hash_to_str(hash:str):
    nlp_en = spacy.load("en_core_web_sm")
    return nlp_en.vocab.strings[hash]

print('Hash codes:')
print(str_to_hash("PERSON"))
print(str_to_hash("Person"))
print(str_to_hash("person"))

print('\nStrings:')
print(hash_to_str(380))
print(hash_to_str(2313063860588076218))
print(hash_to_str(14800503047316267216))

Hash codes:
380
2313063860588076218
14800503047316267216

Strings:
PERSON
Person
person


In [6]:
print('Hash codes:')
print(str_to_hash("PERSON"))
print(str_to_hash("Person"))
print(str_to_hash("person"))

print('\nStrings:')
print(hash_to_str(380))
print(hash_to_str(2313063860588076218))
print(hash_to_str(14800503047316267216))


Hash codes:
380
2313063860588076218
14800503047316267216

Strings:
PERSON
Person
person


#### Exercise - Why does this code throw an error?

```python
import spacy

nlp_en = spacy.load("en_core_web_sm")
nlp_de = spacy.load("de_core_news_sm")
code = nlp_de.vocab.strings["Katze"]
word = nlp_en.vocab.strings[code]
```

The hash for "Katze" is not in the English vocabulary, so the code throws an error.

#### Exercise - And why this works?

```python
import spacy

nlp_en = spacy.load("en_core_web_sm")
nlp_de = spacy.load("de_core_news_sm")
code = nlp_de.vocab.strings["Katze"]

nlp_en.vocab.strings.add("Katze")
word = nlp_en.vocab.strings[code]
```

Here we add the word "Katze" to the English vocabulary, so the code works.

## Inspecting word vectors

SpaCy uses word vectors and with word vectors you can do much more than with just hash codes. 

#### Exercise - "bananas" word vector 1
- Use small `en_core_web_sm` model 
- Print the vector for `"bananas"` using the `token.vector` attribute.
- How long word vector is?

In [7]:
## Task 3:
## Your code here 
nlp_en = spacy.load("en_core_web_sm")
bananas_vector_sm = nlp_en("bananas")[0].vector
print(bananas_vector_sm)
print("Vector length:", len(bananas_vector_sm))

[-0.7758116   1.6605451  -0.39189622 -0.32252163  0.7802841  -0.48605597
  1.371445    1.0607959  -0.3331166  -0.3110752   0.7657112  -0.18480846
 -0.19373728 -0.10193563 -0.05747503  1.1387081  -0.5497641   0.27067196
  0.6271012  -1.2856629  -1.3586068  -0.19643483 -0.14605226  0.26088673
  0.8036362   0.11018939  0.6288574  -0.853708   -0.38757634  0.3829356
 -0.40778196  1.0657507  -0.69960636  0.6153909  -1.099798   -0.75740594
 -0.38754338  1.1433277  -0.01413784  0.01014996  0.6591917  -0.7107733
 -0.9430655  -0.69120514  0.33414608  0.09536967 -0.29030007  0.11057201
 -0.53214085 -1.0528407   0.49351192 -0.6280054  -0.6515591  -0.5959472
  0.02242859  1.2707169   0.9217309   0.4663256   0.53236365  0.75816953
 -0.7602595   0.08518633  0.1791101  -0.8465434   0.02004863  1.1643538
  0.7171087  -0.87006855  0.00917932 -1.0137498  -0.5400826   0.18582445
  0.4986066   0.53966343  0.0084058  -0.718831    0.69892067 -0.7034934
 -1.2734015  -0.08274189 -0.11231536 -0.39930102  0.0849

In [8]:
print(bananas_vector_sm)
print("Vector length:", len(bananas_vector_sm))

[-0.7758116   1.6605451  -0.39189622 -0.32252163  0.7802841  -0.48605597
  1.371445    1.0607959  -0.3331166  -0.3110752   0.7657112  -0.18480846
 -0.19373728 -0.10193563 -0.05747503  1.1387081  -0.5497641   0.27067196
  0.6271012  -1.2856629  -1.3586068  -0.19643483 -0.14605226  0.26088673
  0.8036362   0.11018939  0.6288574  -0.853708   -0.38757634  0.3829356
 -0.40778196  1.0657507  -0.69960636  0.6153909  -1.099798   -0.75740594
 -0.38754338  1.1433277  -0.01413784  0.01014996  0.6591917  -0.7107733
 -0.9430655  -0.69120514  0.33414608  0.09536967 -0.29030007  0.11057201
 -0.53214085 -1.0528407   0.49351192 -0.6280054  -0.6515591  -0.5959472
  0.02242859  1.2707169   0.9217309   0.4663256   0.53236365  0.75816953
 -0.7602595   0.08518633  0.1791101  -0.8465434   0.02004863  1.1643538
  0.7171087  -0.87006855  0.00917932 -1.0137498  -0.5400826   0.18582445
  0.4986066   0.53966343  0.0084058  -0.718831    0.69892067 -0.7034934
 -1.2734015  -0.08274189 -0.11231536 -0.39930102  0.0849

#### Exercise - "bananas" word vector 2
- Repeate same with large model `en_core_web_lg`

In [9]:
## Task 4:
## Your code here 

nlp_en = spacy.load("en_core_web_lg")
bananas_vector_lg = nlp_en("bananas")[0].vector
print(bananas_vector_lg)
print("Vector length:", len(bananas_vector_lg))

[-2.1689e-01 -2.5989e+00 -1.3144e+00  2.2500e+00 -4.6767e-01 -2.0695e+00
 -6.3379e-01 -4.0222e-01 -3.4022e+00 -3.6932e-01 -7.9938e-01 -1.0412e+00
  9.3756e-01  1.6070e+00  8.8330e-01 -2.8483e+00  1.3349e-01 -3.1656e+00
  8.1896e-01 -4.8113e+00  1.5655e+00  1.6665e+00 -4.7081e-01 -1.9475e+00
 -1.1779e+00 -1.3810e+00 -2.0071e+00 -2.1639e-01  9.0609e-01  1.5279e+00
  1.2587e-04 -2.9000e+00  7.6069e-01 -2.2825e+00  1.2495e-02 -1.5653e+00
  2.0052e+00 -1.7747e+00  5.9220e-01 -1.1428e+00 -1.3441e+00  3.4784e-01
  1.7492e+00  1.9086e+00  1.0600e+00  1.2965e+00  4.1431e-01  7.9416e-01
 -1.1277e+00 -1.1403e+00  7.5891e-01 -9.4419e-01  1.4413e+00 -2.2554e+00
  1.6226e-01  3.8901e-01  1.2299e-01  1.1577e+00  1.5524e+00  1.3853e+00
  1.1112e+00  7.5767e-01  3.9431e+00 -2.8506e-01 -2.1645e+00 -1.0862e+00
 -1.4973e+00 -1.2781e+00  2.4643e+00 -1.5886e+00  2.5679e-01  6.4918e-01
  1.6809e-01  5.7693e-01  3.1121e-01 -4.5278e-01 -2.7555e+00 -2.1846e+00
  4.4865e+00  2.7107e-01 -5.3831e-01  8.3013e-01  6

In [10]:
print(bananas_vector_lg)
print("Vector length:", len(bananas_vector_lg))

[-2.1689e-01 -2.5989e+00 -1.3144e+00  2.2500e+00 -4.6767e-01 -2.0695e+00
 -6.3379e-01 -4.0222e-01 -3.4022e+00 -3.6932e-01 -7.9938e-01 -1.0412e+00
  9.3756e-01  1.6070e+00  8.8330e-01 -2.8483e+00  1.3349e-01 -3.1656e+00
  8.1896e-01 -4.8113e+00  1.5655e+00  1.6665e+00 -4.7081e-01 -1.9475e+00
 -1.1779e+00 -1.3810e+00 -2.0071e+00 -2.1639e-01  9.0609e-01  1.5279e+00
  1.2587e-04 -2.9000e+00  7.6069e-01 -2.2825e+00  1.2495e-02 -1.5653e+00
  2.0052e+00 -1.7747e+00  5.9220e-01 -1.1428e+00 -1.3441e+00  3.4784e-01
  1.7492e+00  1.9086e+00  1.0600e+00  1.2965e+00  4.1431e-01  7.9416e-01
 -1.1277e+00 -1.1403e+00  7.5891e-01 -9.4419e-01  1.4413e+00 -2.2554e+00
  1.6226e-01  3.8901e-01  1.2299e-01  1.1577e+00  1.5524e+00  1.3853e+00
  1.1112e+00  7.5767e-01  3.9431e+00 -2.8506e-01 -2.1645e+00 -1.0862e+00
 -1.4973e+00 -1.2781e+00  2.4643e+00 -1.5886e+00  2.5679e-01  6.4918e-01
  1.6809e-01  5.7693e-01  3.1121e-01 -4.5278e-01 -2.7555e+00 -2.1846e+00
  4.4865e+00  2.7107e-01 -5.3831e-01  8.3013e-01  6

## Comparing similarities

In this part, you’ll be using spaCy’s similarity methods to compare Doc, Token and Span objects and get similarity scores.

***Note.*** *SpaCy uses cosine similarity as default.*


In [11]:
nlp = spacy.load("en_core_web_lg")

doc1 = nlp("It's a warm summer day.")
doc2 = nlp("It's sunny outside.")

# Get the similarity of doc1 and doc2
similarity = doc1.similarity(doc2)
print(similarity)

0.8573941945112061


#### Exercise - "bananas" similarity

- Use `en_core_web_lg` model
- Use the `token.similarity` method to compare `bananas` to words:

```python
words = "bananas banana apple apples orange oranges monkey monkeys cat human car space"
```


In [12]:
## Task 5:
## Your code here 
nlp = spacy.load("en_core_web_lg")
bananas = nlp("bananas")[0] # token
words = ["bananas","banana", "apple", "apples", "orange", "oranges", "monkey", "monkeys", "cat", "human", "car", "space"]
bananas_similarities = []
for word in words:
    token = nlp(word)[0]
    similarity = bananas.similarity(token) # compare the similarity of the tokens 
    bananas_similarities.append((word, similarity))
for iii in range(len(bananas_similarities)) :
    print(bananas_similarities[iii])


('bananas', 1.0)
('banana', 0.8712358474731445)
('apple', 0.6073679327964783)
('apples', 0.7323371171951294)
('orange', 0.4965249300003052)
('oranges', 0.6361824870109558)
('monkey', 0.38314178586006165)
('monkeys', 0.43830206990242004)
('cat', 0.2011852264404297)
('human', 0.08909513801336288)
('car', 0.0909561812877655)
('space', 0.08698709309101105)


In [13]:
for iii in range(len(bananas_similarities)) :
    print(bananas_similarities[iii])

('bananas', 1.0)
('banana', 0.8712358474731445)
('apple', 0.6073679327964783)
('apples', 0.7323371171951294)
('orange', 0.4965249300003052)
('oranges', 0.6361824870109558)
('monkey', 0.38314178586006165)
('monkeys', 0.43830206990242004)
('cat', 0.2011852264404297)
('human', 0.08909513801336288)
('car', 0.0909561812877655)
('space', 0.08698709309101105)


#### Exercise - Longer text similarity

- Use `en_core_web_lg` model
- Create function text_similarity(ref, text) that uses `span.similarity` to compare similarities between `ref vs. text`

**Note.** Both texts contains same words, but text2 word order is partly random.

In [14]:
## Task 6:
## Your code here 
def text_similarity(ref, text):
    nlp = spacy.load("en_core_web_lg")
    ref_doc = nlp(ref)
    text_doc = nlp(text)
    similarity = ref_doc.similarity(text_doc)
    return similarity


In [15]:
ref = "Alice likes bananas and apples. She often eats fruits, when she comes from school."
text1 = "Alice likes bananas and apples. She eats fruits all the time, when she comes from work."
text2 = "Alice bananas apples likes and. She eats fruits all the time, when she comes from work."

print('Similarity:', text_similarity(ref, text1))
print('Similarity:', text_similarity(ref, text2))

Similarity: 0.9683893403378143
Similarity: 0.9683893400624004


#### Exercise - Longer text similarity with BLEU

Create similarity comparisons function with BLEU. You get "official" BLEU function with following import. (You may need to install `sacrebleu` library first.)

```python
from sacrebleu.metrics import BLEU
```
https://github.com/mjpost/sacrebleu?tab=readme-ov-file


In [ ]:
## Task 7:
## Your code here 
#!pip install sacrebleu
from sacrebleu.metrics import BLEU

def text_bleu(ref, text):
    ref_list = [ref]  # List of reference sentences
    hyp_list = [text]  # List of hypothesis sentences
    
    bleu = BLEU()
    score = bleu.corpus_score(hyp_list, [ref_list])

    return score

In [17]:
print('BLEU:', text_bleu(ref, text1))
print('BLEU:', text_bleu(ref, text2))

BLEU: BLEU = 54.02 78.9/61.1/47.1/37.5 (BP = 1.000 ratio = 1.118 hyp_len = 19 ref_len = 17)
BLEU: BLEU = 27.60 78.9/33.3/17.6/12.5 (BP = 1.000 ratio = 1.118 hyp_len = 19 ref_len = 17)


expected result:<br>BLEU: BLEU = 54.02 78.9/61.1/47.1/37.5 (BP = 1.000 ratio = 1.118 hyp_len = 19 ref_len = 17) <br>
BLEU: BLEU = 27.60 78.9/33.3/17.6/12.5 (BP = 1.000 ratio = 1.118 hyp_len = 19 ref_len = 17)


## Stop words

Stop words are words that are labeled as words that have no information in text analysis. Therefore these words are often removed before text analysis.

#### Exercise - Stop words

- Count Finnish stop words and print 5 of them
- Count English stop words and print 5 of them

**Hint.** Stop words are python sets. So you need to convert them to lists before you can select first 5 of them. 

In [18]:
## Task 8:
## Your code here 
# count finnish stop words and print 5 of them
nlp = spacy.load("fi_core_news_lg")
stop_words = nlp.Defaults.stop_words
len_stopwords_fi = len(stop_words)
five_stopwords_fi = list(stop_words)[:5]

nlp = spacy.load("en_core_web_lg")
stop_words = nlp.Defaults.stop_words
len_stopwords_en = len(stop_words)
five_stopwords_en = list(stop_words)[:5]
print("FI:", len_stopwords_fi)
print(five_stopwords_fi)
print("EN:", len_stopwords_en)
print(five_stopwords_en)

FI: 822
['pienestä', 'liki', 'älköön', 'vastakkain', 'helposti']
EN: 326
['did', 'latter', 'could', 'nevertheless', 'yourselves']


In [19]:
print("FI:", len_stopwords_fi)
print(five_stopwords_fi)
print("EN:", len_stopwords_en)
print(five_stopwords_en)

FI: 822
['pienestä', 'liki', 'älköön', 'vastakkain', 'helposti']
EN: 326
['did', 'latter', 'could', 'nevertheless', 'yourselves']


expected print: <br>
FI: 822<br>
['menee', 'häntä', 'kuinkaan', 'keille', 'useita']<br>
EN: 326<br>
['elsewhere', 'be', 'for', 'thereupon', 'show']<br>

#### Exercise - Removing stop words

- Use `"en_core_web_lg"` model 
- Create function `` that filters stopwords from the document
- print the original and filtered text

**Hint.** You can find stopwords from document by using `token.is_stop` -tag.

In [20]:
## Task 9:
## Your code here 

def filter_stop_en(text):
    nlp = spacy.load("en_core_web_lg")
    doc = nlp(text)
    filtered_text = [token.text for token in doc if not token.is_stop]
    return " ".join(filtered_text)


In [21]:
alice_sentense = 'Alice like bananas and apples. She often eats fruits, when she come from school.'

# Print the text excluding stop words
print(alice_sentense)
print(filter_stop_en(alice_sentense))

Alice like bananas and apples. She often eats fruits, when she come from school.
Alice like bananas apples . eats fruits , come school .


Expected output :<br>
Alice like bananas and apples. She often eats fruits, when she come from school.<br>
Alice like bananas apples . eats fruits , come school .

#### Exercise - Removing stop words (Finnish)

- Repeat same with Finnish pipeline `fi_core_news_lg` and sentense  
  ```python
  alice_lause = 'Alice pitää banaaneista ja omenoista. Hän syö usein hedelmiä koulusta tullessaan.'
  ```

In [22]:
## Task 10:
## Your code here 

def filter_stop_fi(text):
    nlp = spacy.load("fi_core_news_lg")
    doc = nlp(text)
    filtered_text = [token.text for token in doc if not token.is_stop]
    return " ".join(filtered_text)

In [23]:
alice_lause = 'Alice pitää banaaneista ja omenoista. Hän syö usein hedelmiä koulusta tullessaan.'

# Print the text excluding stop words
print(alice_lause)
print(filter_stop_fi(alice_lause))

Alice pitää banaaneista ja omenoista. Hän syö usein hedelmiä koulusta tullessaan.
Alice pitää banaaneista omenoista . syö hedelmiä koulusta tullessaan .


Expected print:<br>
Alice pitää banaaneista ja omenoista. Hän syö usein hedelmiä koulusta tullessaan. <br>
Alice pitää banaaneista omenoista . syö hedelmiä koulusta tullessaan .

# Reflection
1. What is `"word vector"`? And how many features spaCy `"word vector"` has?
2. What is "bag of words"?
3. What is lemma? What about lemmatization?
4. Why spaCy does not use stemming?
5. Based on what Spacy calculates similaties between sentences?

1. Word vector is a numeric representation of a word in a continuous vector space, which is used to represent words in a way that captures their meaning. SpaCy word vector has 300 features when using the large model and 96 features when using the small model.

2. Bag of words represents a text as a collection (“bag”) of its words, disregarding grammar and word order while retaining the frequency of occurrence.

3. A lemma is the base form of a word, which is used to represent all the inflected forms of the word and lemmatization is the process of converting a word to its base form.

4. SpaCy does not use stemming because probably since it is not accurate enough, as it can produce words that are not real words. Instead, SpaCy uses lemmatization, which is more accurate and produces better results.

5. The calculation between sentences is done by averaging the word embeddings of the words in each sentence and then computing similarity with a similarity measure.
    "The similarity of Doc and Span objects defaults to the average of the token vectors." (source: https://spacy.io/usage/vectors-similarity)<br>
sources:<br> https://spacy.io/ <br> https://stackabuse.com/python-for-nlp-tokenization-stemming-and-lemmatization-with-spacy-library/ <br> https://spacy.io/usage/linguistic-features#vectors-similarity

### Check your answers by running following cell:

In [ ]:
# Do not change this code!

import sys
sys.path.insert(0, '../answers/spacy/')
from spacy2_check import *

print("Results:")
correct = spacy2_check(str_to_hash, hash_to_str, text_similarity, text_bleu, filter_stop_en, filter_stop_fi,
          bananas_vector_sm, bananas_vector_lg, bananas_similarities, 
          len_stopwords_fi, five_stopwords_fi, len_stopwords_en, five_stopwords_en)

print("Correct answers", correct, "/ 13.")

# weidly the bananas_vector_sm is not correct for some random reason ...??? 

Results:
	 'bananas_vector_sm' is not correct. Please check your answer.
Correct answers 12 / 13.


### Nice work! 